### DISCUSSION TITLE
[newbie!] Suitability question

### DISCUSSION THREAD

**Discussion by dietervansteenwegen** [markqvist/Reticulum#309] [Help & Questions] [answered]
---

Disclaimer: I have no experience or background with this stack, and just found out about this after discovering an [open source interface](https://unsigned.io/rnode/) built by someone and subsequently gotten interested. This might be a very stupid question. I DO however have radio experience, interest in new things and enough Python experience to figure things out without solely relying on help from others...

---

I just found this  by accident and I am genuinely intrigued with the whole project. There's a lot of information to go through but I first like to ask if it would be usable for a practical application that I'm working on right now...

I created a system that is sometimes used on events and small festivals. It checks the QR code that people have received in advance by email, verifies that they have a valid invitation, prints out their name/last name/company/... on a badge and logs their attendance. Afterwards, clients often want to have a list of people that [attended/did no show up/registered on-site]. Right now I have separate SQLite databases on each scanning station and have scripting to merge those database entries and process the data afterwards.

Having (near) real-time feedback on a central location would be nice.  I've dabbled a bit with MQTT but in bigger buildings, network connectivity is often flaky so not reliable at all. I was currently looking into setting up a WiFi mesh network or using some LoRa solution but using Reticulum on some dedicated hardware seems much more fun (according to my definition of fun, that is :smile: ) and I'd get to design the hardware as well (bonus points!).

Could Reticulum be a good option to have longer-range communication between the different scanning stations?

Some background:
* The stations are Windows computers, most often things like laptops or [HP ProDesk](https://www.hp.com/be-nl/products/desktops/product-details/2101447420) AIO things.
* My software is written in Python already, which should make compatibility fairly easy.
* **High data throughput or speed is not paramount.** If it takes 30 seconds for a packet of data to be reliably delivered to the "base station" that is perfectly fine. Auto-retry would be nice to have though...
* I need to relay the following information: 
  * person X (identified by a reference, usually about 8 alphanumeric characters) has...
  * checked [in] or [out] at ...
  * scanning post ID Y.
  * So maybe one packet of < 100 bytes, once every .. 20 seconds or so.
* Will be used inside of buildings. Some scanning posts are 5 meters apart, some 100 meters.

I do realize it would be difficult to predict link quality inside of buildings while in different rooms, but I have time to do tests and place relay stations somewhere in between if needed.

Is this a use case where Reticulum could be used (and it is time to start digging in) or should I look elsewhere?

Once I'd have some experience with it, I have some other use cases as well. Obviously ;).

**Comment by dietervansteenwegen** [comment]
After looking further into this, it seems that the fact that the clients run Windows might be an issue. Development is done under Linux, but for the end-use I have to create Windows packages...

Is this correct?

**Reply by cobraPA** [reply]
The messaging client Sideband is available on windows and built with Python and Kivy… it can also run a copy of Reticulum, the networking layer.  Hopefully others will speak up about LoRa connectivity.  The RNode device can bridge to Lora from Bluetooth, and potentially serial/USB, although I haven’t looked to see if it supports all three. (But it’s not a big leap to take your Python client and hook it through USB for Lora on Windows.)

**Reply by dietervansteenwegen** [reply]
Is that true? I can find versions of Sideband for Linux, Mac and Android, but none for Windows. PyPi lists one Sideband package, but that is a logging thing. Am I missing something?

I do realize however that I could likely use the lxmf package as I don't need a user interface, but I'd need to verify that this is working under windows. Neither the PyPi page, nor the author's page mention the compatible operating systems.

**Comment by faragher** [comment, marked as answer]
I love it! Glad to see you here, hope I can give you the answers you need and want. I will try to frame things in what I know of radio, but if there's any ambiguity, please let me know.

I suggest starting with RNodes built on known good hardware, but you can certainly go on from there. I use LilyGo boards marked T3_V1.6.1, which is firmware 2.1 under RNode. Display, LoRa, WiFi, SD (don't know if it's supported by firmware), battery capable, Bluetooth, and multiple antenna options. Ideal development board and a strong contender for "just plain useful." But I sure do get the appeal of making your own hardware. 

Reticulum and Windows is not smooth but not broken. Most of the problems I see are from tty issues with NomadNet. I have used the Python libraries for both raw LoRa output and Reticulum on Windows. I have had more problems on old Raspberry Pis due to the crypto libraries. I can't promise it would work, but I've never had a problem with the base libraries. (if money or time is ultra-tight, I can run through the example suite again to be certain, but I recall every one I used working)

RNS is a network stack, so AFAIK, resending data is on the program itself. However, proof of delivery is possible, and I believe one of the examples. Data rate (as I use it) is around 3200 baud, and if there's a duty cycle limit in your region, that's the responsibility of the software, the radio has no idea the regulatory limits. Given 1,000 bps and 100 bytes per packet, you're still under one second per packet, even with the (tiny) Reticulum overhead.

The question of range is, of course, tough. At 10dBm (max on the LoRa chipset used is 17) I would consider 1,000 feet from an indoor base station to a radio inside a vehicle, as reliable through light suburban and woodland. I've had 2,300 feet of residential, woodland, and even terrain come through, but that was with > 50% dropped packets. So, a hundred meters in a warehouse? Probably (multi-path issues aside). A convention hall? Depends on the walls, but likely. Heavy construction? May be at the limits. Of course, the answer there is to just add a node! Free space with line of sight is almost a gimme at that range.

The important thing to note here is the sensitivity of these devices is astounding. We're talking -130+ dB and well beneath the noise floor. I had an SNR of -11.25 dB come through once! 

If you spend the time to get to know the system, it can probably do what you ask of it. The one thing to consider, however, is the RNodes really aren't standalone at this time. Basically any SBC or Bluetooth enabled phone can run one, but at this time they can't simply stand alone. They're essentially a serial modem. 

I'm afraid so far my focus has been on the radio itself and some networking utility, but I don't fully understand the stack at this time. I may be wrong about specifics. but I think it's a viable option.

**Reply by dietervansteenwegen** [reply]
Awesome, thanks.

I'll order a couple of [these](https://www.tinytronics.nl/shop/nl/development-boards/microcontroller-boards/met-lora/lilygo-ttgo-t3-lora32-433mhz-v1.6.1-esp32), expecting these are the boards you recommend. They don't seem to be mentioned in the [list of supported hardware](https://reticulum.network/manual/hardware.html#supported-boards) for the auto-installer, but if you've used them and recommend them, that's where I'll start. As said, I've just started reading up on this yesterday night with no prior experience (and a head cold so I'm a bit slow at the moment), so things are still overwhelming. 

I might end up designing my own hardware (since I would like to include additional functionality if I need an external component to the laptop anyway), but these seem very nice to start testing with.

```text
We're talking -130+ dB and well beneath the noise floor. I had an SNR of -11.25 dB come through once!
```

That is nice to hear. Given the cost of these things, adding a node as a repeater is a no-brainer. 

```text
The one thing to consider, however, is the RNodes really aren't standalone at this time. 
Basically any SBC or Bluetooth enabled phone can run one, but at this time they can't simply stand alone. 
They're essentially a serial modem.
```

Not sure how to interpret this as [this page from Unsigned](https://unsigned.io/hardware/RNode.html) seems to claim differently, offering a stand-alone piece of hardware. But I might be wrong, obviously. In any case, building a case with a Beaglebone or any other SBC shouldn't be too bad. (although, if something is available commercially that's likely much cheaper in the long run)

**Reply by faragher** [reply]
> I'll order a couple of [these](https://www.tinytronics.nl/shop/nl/development-boards/microcontroller-boards/met-lora/lilygo-ttgo-t3-lora32-433mhz-v1.6.1-esp32), expecting these are the boards you recommend. They don't seem to be mentioned in the [list of supported hardware](https://reticulum.network/manual/hardware.html#supported-boards) for the auto-installer, but if you've used them and recommend them, that's where I'll start. As said, I've just started reading up on this yesterday night with no prior experience (and a head cold so I'm a bit slow at the moment), so things are still overwhelming.

Those appear to be the board I am using. The auto-installer nomenclature is a little different. It defines this board as the LilyGO LoRa32 v2.1 and that's what you'd like to select with the rnodeconf program as well.
 
> I might end up designing my own hardware (since I would like to include additional functionality if I need an external component to the laptop anyway), but these seem very nice to start testing with.

Let us know if it's open sourced or you put it up for sale. Personally, I'd love something stand-alone with a GPS chip, but applications can vary so wildly.


> ```
> The one thing to consider, however, is the RNodes really aren't standalone at this time. 
> Basically any SBC or Bluetooth enabled phone can run one, but at this time they can't simply stand alone. 
> They're essentially a serial modem.
> ```
> 
> Not sure how to interpret this as [this page from Unsigned](https://unsigned.io/hardware/RNode.html) seems to claim differently, offering a stand-alone piece of hardware. But I might be wrong, obviously. In any case, building a case with a Beaglebone or any other SBC shouldn't be too bad. (although, if something is available commercially that's likely much cheaper in the long run)

I can only speak to my understanding of the system. There's been some demand for a stand-alone Reticulum node, but that's embedded C, not Python. (although some Python can be used on some microcontrollers, but I know nothing about the specifics, so I can't speculate)  As far as I know, the RNode is just a modem for packet radio. Since I can't get my hands on Pi for reasonable prices anymore, I've picked up some Le Potatoes for testing, but I wish I could find some Pi Zeros or similar boards to test. The OTG USB connector may cost more than the SBC at that point, but it would be nice to use a minimal amount of hardware. If you find a Beagleboard that works well, let us know. The Pocket is more than these Potatoes, but that form factor would make for a real pocket radio system.
